File: 1081. smallest-subsequence-of-distinct-characters.java
import java.util.Stack;

class Solution {
    public String smallestSubsequence(String s) {
        int[] freq = new int[27];
        boolean[] seen = new boolean[27];
        Stack<Character> stack = new Stack<>();

        for (int i = 0; i < s.length(); i++)
            freq[s.charAt(i) & 31]++;

        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            int x = c & 31;
            freq[x]--;

            if (seen[x])
                continue;

            while (!stack.isEmpty()) {
                if (stack.peek() <= c)
                    break;

                if (freq[stack.peek() & 31] == 0)
                    break;

                seen[stack.peek() & 31] = false;
                stack.pop();
            }

            stack.push(c);
            seen[x] = true;
        }

        StringBuilder res = new StringBuilder();

        for (char c : stack)
            res.append(c);

        return res.toString();
    }

    public static void main(String[] args) {
        Solution sol = new Solution();

        // Example 1
        String s1 = "bcabc";
        System.out.println(sol.smallestSubsequence(s1)); // abc

        // Example 2
        String s2 = "cbacdcbc";
        System.out.println(sol.smallestSubsequence(s2)); // acdb
    }
}

File: 1081. smallest-subsequence-of-distinct-characters.py
from collections import Counter


class Solution:
    def smallestSubsequence(self, s: str) -> str:
        freq = Counter(s)
        seen = set()
        stack = []

        for c in s:
            freq[c] -= 1
            if c in seen:
                continue

            while stack and stack[-1] > c and freq[stack[-1]]:
                seen.remove(stack.pop())

            stack.append(c)
            seen.add(c)

        return "".join(stack)


if __name__ == "__main__":
    sol = Solution()

    # Example 1
    s = "bcabc"
    print(sol.smallestSubsequence(s))  # "abc"

    # Example 2
    s = "cbacdcbc"
    print(sol.smallestSubsequence(s))  # "acdb"